### PR TITLE
RavenDB-17243: Implement native support for nulls

### PR DESCRIPTION
**Pull request by redknightlois**
After some working around I was able to piggyback the null support in an space efficient way and with as less as possible disruption into the format. More complex format means that the decoding phase would be more costly to execute. Therefore, keeping the format as simple as possible was a requirement. 

Current implementation supports handling null at every property level (albeit some options are not implemented yet on this draft PR). 

cc @maciejaszyk check if there is something else that's required to use native nulls instead of null strings.

### CODE REVIEW

File: src/Corax/IndexEntryBuilder.cs [hunk 1]
```diff
@@ -256,14 +229,41 @@ public void FinishWritingEnumerable(int metadataLocation, int dataLocation, int
...
+            {
+                // Copy the null stream.
+                new ReadOnlySpan<byte>(nullStream, nullBitStreamSize * sizeof(long))
+                    .CopyTo(_buffer.Slice(dataLocation));
+
+                dataLocation += nullBitStreamSize;
+
+                // Signal that we will have to deal with the nulls.
+                _buffer[indexEntryFieldLocation] |= (byte)IndexEntryFieldType.HasNull;
+            }
```
Comment (ayende) [inline src/Corax/IndexEntryBuilder.cs]: This is duplicated code, no?

File: src/Corax/IndexEntryBuilder.cs [hunk 2]
```diff
@@ -305,7 +308,34 @@ public void Write(int field, IReadOnlySpanEnumerator values, ReadOnlySpan<long>
...
```
Comment (ayende) [inline src/Corax/IndexEntryBuilder.cs]: Duplicated code

File: src/Corax/IndexEntryBuilder.cs [hunk 3]
```diff
@@ -473,51 +504,85 @@ internal IndexEntryFieldIterator(IndexEntryFieldType type)
...
         }
 
+        public bool IsNull
+        {
+            get
+            {
+                if (!Type.HasFlag(IndexEntryFieldType.HasNull))
+                    return false;
+
+                throw new NotImplementedException();
```
Comment (ayende) [inline src/Corax/IndexEntryBuilder.cs]: ??

File: src/Corax/IndexEntryBuilder.cs [hunk 4]
```diff
@@ -842,12 +916,28 @@ public bool Read(int field, out IndexEntryFieldType type, out Span<byte> value,
                 }
                 else if ((type & IndexEntryFieldType.Raw) != 0)
                 {
+                    if (type.HasFlag(IndexEntryFieldType.HasNull))
+                    {
+                        var spanTableOffset = MemoryMarshal.Read<int>(_buffer[intOffset..]);
+                        byte* nullTablePtr = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(_buffer));
+                        if (PtrBitVector.GetBitInPointer(nullTablePtr + spanTableOffset, elementIdx) == true)
+                            goto HasNull;
+                    }
```
Comment (ayende) [inline src/Corax/IndexEntryBuilder.cs:954]: No need to skip the null table here?

File: src/Corax/IndexEntryBuilder.cs [hunk 5]
```diff
@@ -842,12 +916,28 @@ public bool Read(int field, out IndexEntryFieldType type, out Span<byte> value,
...
                 }
                 else if ((type & IndexEntryFieldType.Tuple) != 0)
                 {
+                    if (type.HasFlag(IndexEntryFieldType.HasNull))
+                    {
+                        var spanTableOffset = MemoryMarshal.Read<int>(_buffer[intOffset..]);
+                        byte* nullTablePtr = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(_buffer));
+                        if (PtrBitVector.GetBitInPointer(nullTablePtr + spanTableOffset, elementIdx) == true)
+                            goto HasNull;
+                    }
```
Comment (ayende) [inline src/Corax/IndexEntryBuilder.cs:968]: No need to skip the null table here?

File: test/FastTests/Corax/IndexEntryWriter.cs [hunk 6]
```diff
@@ -31,7 +31,15 @@ public StringArrayIterator(string[] values)
 
             public int Length => _values.Length;
 
-            public ReadOnlySpan<byte> this[int i] => Encoding.UTF8.GetBytes(_values[i]);
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
```
Comment (ayende) [inline test/FastTests/Corax/IndexEntryWriter.cs]: Do we need it, the `_values` already does this, no?

File: test/FastTests/Corax/IndexSearcher.cs [hunk 7]
```diff
@@ -52,6 +52,14 @@ public StringArrayIterator(IEnumerable<string> values)
 
             public int Length => _values.Length;
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
+
```
Comment (ayende) [inline test/FastTests/Corax/IndexSearcher.cs]: Do we need it, the `_values` already does this, no?

File: test/FastTests/Sparrow/Encoder3GramTests.cs [hunk 8]
```diff
@@ -33,6 +33,14 @@ private struct StringKeys : IReadOnlySpanEnumerator, ISpanEnumerator
 
             public int Length => _values.Length;
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
```
Comment (ayende) [inline test/FastTests/Sparrow/Encoder3GramTests.cs]: Do we need it, the `_values` already does this, no?

File: test/Tryouts/CoraxEnron.cs [hunk 9]
```diff
@@ -301,6 +301,14 @@ public StringArrayIterator(IEnumerable<string> values)
 
             public int Length => _values.Length;
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
```
Comment (ayende) [inline test/Tryouts/CoraxEnron.cs]: Do we need it, the `_values` already does this, no?

File: test/Tryouts/Program.cs [hunk 10]
```diff
@@ -90,6 +90,14 @@ public StringArrayIterator(string[] values)
 
             public int Length => _values.Length;
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
+
```
Comment (ayende) [inline test/Tryouts/Program.cs]: Do we need it, the `_values` already does this, no?

**Review by ayende — COMMENTED** [review]
If I understand correctly, null handling is using:

* An additional flag on the entries (`HasNull`)
* If that flag is set, we add a bitfield that mark null values for the particular fields?

How would you then handle a scenario an array with null in it?

`{ Stages: [ "stages/1", null, "stages/3"] }`

And then asking for: `Stages = null` ?

File: src/Corax/IndexEntryBuilder.cs (hunk 2 above)
Comment (redknightlois) [reply src/Corax/IndexEntryBuilder.cs]: Yes, for now it will stay like that.

**Comment by redknightlois** [comment]
> If I understand correctly, null handling is using:
> 
> * An additional flag on the entries (`HasNull`)
> * If that flag is set, we add a bitfield that mark null values for the particular fields?
> 
> How would you then handle a scenario an array with null in it?
> 
> `{ Stages: [ "stages/1", null, "stages/3"] }`
> 
> And then asking for: `Stages = null` ?

There are 2 different ways to access that. The `Stages = null` actually returns null for the item entirely. It will be written differently than the case of  `{ Stages: [ "stages/1", null, "stages/3"] }` . At the protocol level, these are entirely different formats. One is a list, the other is a plain null on the `Stages` field.

File: test/FastTests/Sparrow/Encoder3GramTests.cs (hunk 8 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Encoder3GramTests.cs]: Not really

File: src/Corax/IndexEntryBuilder.cs (hunk 3 above)
Comment (redknightlois) [reply src/Corax/IndexEntryBuilder.cs]: Still not implemented... I use that to be able to kill the cases where I use that feature :)

File: src/Corax/IndexEntryBuilder.cs (hunk 4 above)
Comment (redknightlois) [reply src/Corax/IndexEntryBuilder.cs:954]: Null table does not exist if there are no potential nulls stored. So no need to skip. This is a pay-for-usage type of thing... No NULL, no pay.

**Comment by ayende** [comment]
> 

`Stages = null` means - currently - a `Stages` field hat has a `null` value in it. Not just set to null, but also an array that has a null value.

**Comment by redknightlois** [comment]
> > 
> 
> `Stages = null` means - currently - a `Stages` field hat has a `null` value in it. Not just set to null, but also an array that has a null value.

Yes. Those are two different. The case where 'Stages' is 'null' has the flag set up as null. Flags is equal to 0. The case of the array containing a null value is a HasNull and with the internal marker for nulls.

**Comment by arekpalinski** [comment]
test this please

**Review by arekpalinski — APPROVED** [review]

**Comment by redknightlois** [comment]
The change to use `ushort` instead of `byte` as type header is done. Required standardizing the access.

**Comment by arekpalinski** [comment]
@redknightlois can you check `SlowTests.Corax.RavenDB_18377.SlowTests.Corax.RavenDB_18377.ArrayInIndex(options: DatabaseMode = Single , SearchEngineMode = Corax)` test?

**Comment by redknightlois** [comment]
@arekpalinski Either we fixed (inadvertedly) a bug when working on NULL support OR it was already wrong. The conditions were inverted between the static index and the auto index. There is absolutely no reason why they would be returning Empty and the Single and both pass in green.

**Comment by arekpalinski** [comment]
It was a corax specific test so I believe we just put the wrong conditions which was passing at that time, even though they were not valid.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexWriteOperation.cs [hunk 11]
```diff
@@ -66,9 +66,6 @@ public override void Commit(IndexingStatsScope stats)
 
             using (Stats.ConvertStats.Start())
                 data = _converter.SetDocumentFields(key, sourceDocumentId, document, indexContext, out lowerId, _buffer.ToSpan());
-
-            if (data.IsEmpty)
-                return;
             
             using (Stats.AddStats.Start())
                 _indexWriter.Index(lowerId, data, _knownFields);
```
Comment (redknightlois) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexWriteOperation.cs]: @maciejaszyk Is there any way I can know here or before if the document is actually an implicit null (and/or explicit)?

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexWriteOperation.cs]: @redknightlois 
In `CoraxDocumentConverterBase` we have something in constructor:
image

**Comment by redknightlois** [comment]
I am enabling tests to be run on Corax with the objective of finding edge cases. So far, found 1.

**Comment by arekpalinski** [comment]
Seems like there is just one failure - https://jenkins.hibernatingrhinos.com/job/PR_Tests/1903/testReport/junit/SlowTests.Bugs.Queries/QueryingOnEqualToNull/SlowTests_Bugs_Queries_QueryingOnEqualToNull_QueryingOnEqNotNull_options___DatabaseMode___Single___SearchEngineMode___Corax_/

**Comment by redknightlois** [comment]
@maciejaszyk Can you look into the `QueryingOnEqNotNull` failing test? It seems to me that it is really unrelated to #14158 and more related to case sensitivity and how we treat the `id()` field. Missing some context to fix it. If you need we can have a call in my morning. 

@arekpalinski If all the slow tests pass, coordinate with Maciej as this PR is a dependency for #14164.

**Comment by arekpalinski** [comment]
test this please

File: src/Corax/Queries/UnaryMatch.Ordinal.cs [hunk 12]
```diff
@@ -191,129 +435,147 @@ private static int FillFuncNumerical<TComparer>(ref UnaryMatch<TInner, TValueTyp
...
             }
             else
             {
                 return new UnaryMatch<TInner, TValueType>(
                     in inner, UnaryMatchOperation.NotEquals,
                     searcher, fieldId, value,
-                    &FillFuncNumerical<NotEqualsMatchComparer>, &AndWith,
-                    inner.Count, inner.Confidence.Min(QueryCountConfidence.Normal), distinct, take);
+                    mode == UnaryMatchOperationMode.Any ? &FillFuncNumerical<EqualsMatchComparer> : &FillFuncNumerical<EqualsMatchComparer>, &AndWith,
+                    inner.Count, inner.Confidence.Min(QueryCountConfidence.Normal), mode, true, take: take);
```
Comment (maciejaszyk) [inline src/Corax/Queries/UnaryMatch.Ordinal.cs]: You are passing `EqualsMatchComparer` in `NotEqualMatch` ;) 
@redknightlois

Comment (maciejaszyk) [reply src/Corax/Queries/UnaryMatch.Ordinal.cs]: This is about `QueryingOnEqNotNull`

**Comment by maciejaszyk** [comment]
Some other tests failed after the fix - https://github.com/ravendb/ravendb/pull/14244
Probably this is related to UnaryMatch logic.

**Comment by maciejaszyk** [comment]
BTW: 
I think we should skip the `null` value at `ExistsMatch`

File: src/Corax/Queries/UnaryMatch.Ordinal.cs (hunk 12 above)
Comment (redknightlois) [reply src/Corax/Queries/UnaryMatch.Ordinal.cs]: I am passing `EqualsMatchComparer` but I am also forcing `distinct = true` ... this is one of the side effects of `null` you have to be very careful with the `Any` and `All` distinction. It is not symmetric as it was before... maybe the failed ones are related to `FillFuncNumerical` as I am seeing that they have no null handling.

**Comment by redknightlois** [comment]
@maciejaszyk Check `QueryingOnEqNotNull` it seems it is not even related to NULLs it just surfaced because of it. The problem (at least it seems to me) is one of casing of `id()` fields.
